src/modal_analysis.cu

```diff
@@ -479,57 +524,51 @@ void MODAL_ANALYSIS::preprocess(char *input_dir, Atom *atom)
         for (int i=1; i<first_mode; i++) { getline(eigfile,val); }
         for (int j=0; j<num_modes; j++) //modes
         {
-            set_eigmode(j, eigfile, eig_x);
-            set_eigmode(j, eigfile, eig_y);
-            set_eigmode(j, eigfile, eig_z);
+            set_eigmode(j, eigfile, eigx);
+            set_eigmode(j, eigfile, eigy);
+            set_eigmode(j, eigfile, eigz);
         }
         eigfile.close();
 
         // Allocate intermediate vector
-        size_t mv_n_size = sizeof(float) * num_participating;
-        CHECK(cudaMallocManaged((void **)&mv_x, mv_n_size));
-        CHECK(cudaMallocManaged((void **)&mv_y, mv_n_size));
-        CHECK(cudaMallocManaged((void **)&mv_z, mv_n_size));
+        size_t mv_size = num_participating;
+        mvx.resize(mv_size, Memory_Type::managed);
+        mvy.resize(mv_size, Memory_Type::managed);
+        mvz.resize(mv_size, Memory_Type::managed);
 
         // Allocate modal velocities
-        size_t xdot_size = sizeof(float) * num_modes;
-        CHECK(cudaMallocManaged((void **)&xdot_x, xdot_size));
-        CHECK(cudaMallocManaged((void **)&xdot_y, xdot_size));
-        CHECK(cudaMallocManaged((void **)&xdot_z, xdot_size));
+        size_t xdot_size = num_modes;
+        xdotx.resize(xdot_size, Memory_Type::managed);
+        xdoty.resize(xdot_size, Memory_Type::managed);
+        xdotz.resize(xdot_size, Memory_Type::managed);
 
         // Allocate modal measured quantities
-        size_t jmxyz_size = sizeof(float) * num_modes*3;
-        CHECK(cudaMallocManaged((void **)&jmx,jmxyz_size));
-        CHECK(cudaMallocManaged((void **)&jmy,jmxyz_size));
-        CHECK(cudaMallocManaged((void **)&jmz,jmxyz_size));
+        size_t jmxyz_size = num_modes*3;
+        jmx.resize(jmxyz_size, Memory_Type::managed);
+        jmy.resize(jmxyz_size, Memory_Type::managed);
+        jmz.resize(jmxyz_size, Memory_Type::managed);
 
         num_heat_stored = num_modes*NUM_OF_HEAT_COMPONENTS;
-        size_t jm_size = sizeof(float) * num_heat_stored;
-        CHECK(cudaMallocManaged((void **)&jm,jm_size));
+        size_t jm_size = num_heat_stored;
+        jm.resize(jm_size, 0.0f, Memory_Type::managed);
```

Comment:
Here is another possible reason for slow-down. In the original version, the unified memory (managed memory) was first touched by GPU using the kernel function gpu_reset_data(). In the updated version, the unified memory was first touched by the CPU based on the function call resize(jm_size, 0.0f, Memory_Type::managed). The updated approach can lead to more page-fault conditions and cost some time. I will update the resize() function to allow it accepting a more optional argument that allow you choose where the data are initialized (CPU or GPU).

A final solution is to make choices for Windows and Linux code paths and add some optimization code for the Linux path. For example, the prefetch() function is very useful when using unified memory in Linux with CC>=6.0, but unfortunately, it is not supported in Windows for any GPU.

Reply:
Can you tried to switch back to the original kernel function to initialize the data and check if the speed is affected by the choice?

Reply:
You are right that the new initialization will cause some page-faults. I switched back to the original kernel function, but it did not make a difference. Perhaps the array is too small to matter. 

At one point, I did try pre-fetching all of the GPU-relevant data in the preprocessing function. Unfortunately, the improvements were barely perceptible and the code was harder to read. I think this makes sense because most data is only loaded to the GPU once and then remains there until it is freed. Page-faults only occur for the first time the method is called.

For a sanity check, I revisited the code from #55 to double check the speeds. This time, the code runs slower than I reported in the pull request. It seems to be on par with the new code using GPU_Vector (after fixing the eig_size error you pointed out). After running 20+ tests, I find that there are some outlier simulations (fast and slow) that can mess with averages. At this point, I am satisfied and believe there is no real speed difference after updating to GPU_Vector.